internal/lock.go

```diff
@@ -49,20 +57,92 @@ func NewLock(path string, newOk bool) (*Lock, error) {
 	return &Lock{path: path, conf: conf}, nil
 }
 
-func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string) error {
+func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string, cacheURL string) error {
 	for _, u := range paths {
 		if l.Contains(u) {
 			return fmt.Errorf("resource '%s' is already present", u)
 		}
 	}
-	r, err := NewResourceFromUrl(paths, algo, tags, filename)
+	r, err := NewResourceFromUrl(paths, algo, tags, filename, cacheURL)
 	if err != nil {
 		return err
 	}
+
+	// If cache URL is provided, handles Artifactory upload
+	if cacheURL != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token == "" {
+			return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+		}
+	}
+	// Get file again for upload
+	path, err := GetUrltoTempFile(paths[0], context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to get file for cache: %s", err)
+	}
+	defer os.Remove(path)
+
+	// Upload to Artifactory using hash as filename
+	err = uploadToArtifactory(path, cacheURL, r.Integrity)
+	if err != nil {
+		return fmt.Errorf("failed to upload to cache: %v", err)
+	}
+
 	l.conf.Resource = append(l.conf.Resource, *r)
 	return nil
 }
 
+func uploadToArtifactory(filePath, cacheURL, integrity string) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+	}
+
+	// Read file
+	fileContent, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to read file: %v", err)
+	}
+
+	// Get hash from integrity string
+	h := strings.TrimPrefix(integrity, "sha256-")
+	h = strings.TrimRight(h, "=")
+	padding := len(h) % 4
+	if padding != 0 {
+		h += strings.Repeat("=", 4-padding)
+	}
+
+	hashBytes, err := base64.StdEncoding.DecodeString(h)
+	if err != nil {
+		return fmt.Errorf("failed to decode hash: %v", err)
+	}
+
+	hexHash := hex.EncodeToString(hashBytes)
+	artifactoryURL := fmt.Sprintf("%s/%s", cacheURL, hexHash)
+
+	// Create upload request
+	req, err := http.NewRequest("PUT", artifactoryURL, bytes.NewReader(fileContent))
```

Comment:
As discussed, let's use `github.com/carlmjohnson/requests` for non-test http interactions, like the rest of the code does.